sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala

```diff
@@ -46,12 +57,17 @@ object Literal {
     case s: String => Literal(UTF8String.fromString(s), StringType)
     case b: Boolean => Literal(b, BooleanType)
     case d: BigDecimal => Literal(Decimal(d), DecimalType(Math.max(d.precision, d.scale), d.scale))
-    case d: java.math.BigDecimal =>
+    case d: JavaBigDecimal =>
       Literal(Decimal(d), DecimalType(Math.max(d.precision, d.scale), d.scale()))
     case d: Decimal => Literal(d, DecimalType(Math.max(d.precision, d.scale), d.scale))
     case t: Timestamp => Literal(DateTimeUtils.fromJavaTimestamp(t), TimestampType)
     case d: Date => Literal(DateTimeUtils.fromJavaDate(d), DateType)
     case a: Array[Byte] => Literal(a, BinaryType)
+    case a: Array[_] =>
```

Comment:
Since `Literal#apply` hold no `TypeTag`, seems we cannot `dataTypeFor` in `case a: Array[_] =>`.
Concretely, I put your code in `case a: Array[_] =>`, I got a compiler error;
```
ror] /Users/maropu/IdeaProjects/spark/spark-master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala:73: type mismatch;
[error]  found   : reflect.runtime.universe.Type
[error]  required: org.apache.spark.sql.catalyst.ScalaReflection.universe.TypeTag[?]
[error]       val dataType = ScalaReflection.dataTypeFor(rm.classSymbol(a.getClass).selfType)
```